add sample (#24767)
Azure/azure-sdk-for-python: sdk/identity/azure-identity/samples/azure-aad-auth-with-redis-py.md

## Azure Cache for Redis: Azure AD with redis client library

### Table of contents

> 

Reviewer: Add Pre-req section with following content as idiomatically applicable:

- Configuration of Role and Role Assignments is required before using the sample code in this document.
- **Dependency Requirements:**
   ```xml
 
   ```

- [Authenticate with Azure AD - Hello World](#authenticate-with-azure-ad-hello-world)

> - [Authenticate with Azure AD - Handle Re-authentication](#authenticate-with-azure-ad-handle-re-authentication)

Reviewer: Add Troubleshooting section as follows:
- [Troubleshooting](#troubleshooting)

Place the TBS section at the bottom of doc.
Add the following content in the TBS section

#### Troubleshooting

##### Invalid Username Password Pair Error
In this error scenario, the username provided and the access token used as password are not compatible.
To mitigate this error, ensure that:
* On Portal, Under your `Redis Cache Resource` -> RBAC Rules, you've assigned the required role to your user/service principal identity.
* On Portal, Under your `Redis Cache Resource` -> Advanced settings -> AAD Access Authorization box is checked/enabled, if not enable it and press the Save button.

##### Permissions not granted / NOPERM Error
In this error scenario, the authentication was successful, but your registered user/service principal is not granted the RBAC permission to perform the action.
To mitigate this error, ensure that:
* On Portal, Under your `Redis Cache Resource` -> RBAC Rules, you've assigned the appropriate role (Owner, Contributor, Reader) to your user/service principal identity.
* In the event you're using a custom role, then ensure the permissions granted under your custom role include the one required for your target action.

#### Samples guidance

Familiarity with the [redis](https://pypi.org/project/redis/) and [azure-identity for Python](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/identity/azure-identity) client libraries is assumed.

[Authenticate with Azure AD - Hello World](#authenticate-with-azure-ad-hello-world)
This sample is recommended for users getting started to use Azure AD authentication with Azure Cache for Redis.

[Authenticate with Azure AD - Handle Re-Authentication](#authenticate-with-azure-ad-handle-re-authentication)
This sample is recommended to users looking to build long-running applications that would like to handle re-authenticating with Azure AD upon token expiry.

> **Note:** The below sample uses the Azure Identity library's `ClientSecretCredential`. The credential can be replaced with any of the other `TokenCredential` implementations offered by the azure-identity library.

Reviewer 2: ```suggestion
**Note:** The below sample uses the Azure Identity library's `DefaultAzureCredential`. The credential can be replaced with any of the other `TokenCredential` implementations offered by the azure-identity library.
```

> **Note:** The below sample uses the Azure Identity library's `DefaultAzureCredential`. The credential can be replaced with any of the other `TokenCredential` implementations offered by the azure-identity library.

Reviewer: Remove this section.

#### Authenticate with Azure AD: hello world

This sample is intended to assist in authenticating with Azure AD via the redis client library. It focuses on displaying the logic required to fetch an Azure AD access token and to use it as password when setting up the redis instance.

##### Migration guidance

When migrating your existing application code, you need to replace the password input with the Azure AD token.
Integrate the logic in your application code to fetch an Azure AD access token via the azure-identity library as shown below and replace it with the password configuring/retrieving logic in your application code.

```py
import redis
from azure.identity import DefaultAzureCredential

> scope = "https://*.cacheinfra.windows.net:10225/appid/.default"  # The scope will be changed for AAD Public Preview

Reviewer 2: I'm a bit out of the loop, so the comment on this line doesn't make much sense to me. Does this mean that the displayed scope is what needs to be used today, but that AAD will be changing the scope as part of a public preview?

Author: Right. This was requested by the feature team to explicitly point out this scope is for temporary use. They will change it for public review.

host = ""  # Required

> port = 6380  # Required

Reviewer 2: Is port 6380 the required port, or is this just an example? If it has to be this port specifically, I think saying so in the comment would be helpful

Author: 6380 is the port for SSL connection. Feature team asked us to use the secure port in our demo.

user_name = ""  # Required

def hello_world():
    cred = DefaultAzureCredential()
    token = cred.get_token(scope)
    r = redis.Redis(host=host,
                    port=port,

>                     ssl=True,

Reviewer: nit: callout ssl connection required.

Reviewer: the comment got applied below, but missed here I think.

                    username=user_name,
                    password=token.token,
                    decode_responses=True)
    r.set("Az:key1", "value1")
    t = r.get("Az:key1")
    print(t)

if __name__ == '__main__':
    hello_world()
```

> 

Reviewer: After Hello World sample, 
add this:

##### Supported Token Credentials for Azure AD Authentication
**Note:** The samples in this doc use the Azure Identity library's `DefaultAzureCredential` to fetch Azure AD Access Token. The other supported `TokenCredential` implementations that can be used from [Azure Identity for Java](https://docs.microsoft.com/azure/developer/java/sdk/identity) are as follows:
* [Client Certificate Credential](https://docs.microsoft.com/azure/developer/java/sdk/identity-service-principal-auth#client-certificate-credential)
* [Client Secret Credential](https://docs.microsoft.com/azure/developer/java/sdk/identity-service-principal-auth#client-secret-credential)
* [Managed Identity Credential](https://docs.microsoft.com/azure/developer/java/sdk/identity-azure-hosted-auth#managed-identity-credential)
* [Username Password Credential](https://docs.microsoft.com/azure/developer/java/sdk/identity-user-auth#username-password-credential)
* [Azure CLI Credential](https://docs.microsoft.com/azure/developer/java/sdk/identity-dev-env-auth#azure-cli-credential)
* [Interactive Browser Credential](https://docs.microsoft.com/azure/developer/java/sdk/identity-user-auth#interactive-browser-credential)
* [Device Code Credential](https://docs.microsoft.com/azure/developer/java/sdk/identity-user-auth#device-code-credential)

Reviewer: Note: Update Hyperlinks for Python Accordingly.

##### Migration guidance

When migrating your existing application code, you need to replace the password input with the Azure AD token.
Integrate the logic in your application code to fetch an Azure AD access token via the azure-identity library as shown below and replace it with the password configuring/retrieving logic in your application code.

```py
import logging
import redis
from azure.identity import DefaultAzureCredential

scope = "https://*.cacheinfra.windows.net:10225/appid/.default"  # The scope will be changed for AAD Public Preview
host = ""  # Required
port = 6380  # Required
user_name = ""  # Required

def re_authentication():
    _LOGGER = logging.getLogger(__name__)
    cred = DefaultAzureCredential()
    token = cred.get_token(scope)
    r = redis.Redis(host=host,
                    port=port,
                    ssl=True,
                    username=user_name,
                    password=token.token,
                    decode_responses=True)
    max_retry = 3
    for index in range(max_retry):
        try:
            r.set("Az:key1", "value1")
            t = r.get("Az:key1")
            print(t)
            break

>         except redis.ConnectionError as ex:

Reviewer 2: nit: it doesn't look like `ex` is used

## Azure Cache for Redis: Azure AD with redis client library

### Table of contents

> - [Dependency requirements](#dependency-requirements)

Reviewer: Replace this with 
`- [Prerequisites](#prerequisites)`

Add the section as follows on top of document:

#### Prerequisites
-  Configuration of Role and Role Assignments is required before using the sample code in this document.
 - **Dependency Requirements:**
    - [redis](https://pypi.org/project/redis/)
    - [azure-identity for Python](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/identity/azure-identity)

## Azure Cache for Redis: Azure AD with redis client library

### Table of contents

- [Prerequisites](#prerequisites)
- [Samples guidance](#samples-guidance)
- [Authenticate with Azure AD - Hello World](#authenticate-with-azure-ad-hello-world)
- [Authenticate with Azure AD - Handle Re-authentication](#authenticate-with-azure-ad-handle-re-authentication)
- [Troubleshooting](#troubleshooting)

#### Prerequisites

> - Configuration of Role and Role Assignments is required before using the sample code in this document.

Reviewer 3: I left this same feedback for Vinay in his PR. We should link to a doc that explains how to configure roles and role assignments. Since it's a prereq, we don't want customers to be blocked if they can't get past that.

Author: @scottaddie Talked to the feature team. The doc will not be ready before our release which is on 8/9 and they would like us to merge the PR w/o the link and add the link when it is ready.

Do you have concerns?

Reviewer 3: No concerns. We can update later.

##### Migration guidance

When migrating your existing application code, you need to replace the password input with the Azure AD token.
Integrate the logic in your application code to fetch an Azure AD access token via the azure-identity library as shown below and replace it with the password configuring/retrieving logic in your application code.

```py
import redis
from azure.identity import DefaultAzureCredential

scope = "https://*.cacheinfra.windows.net:10225/appid/.default"  # The scope will be changed for AAD Public Preview
host = ""  # Required
port = 6380  # Required
user_name = ""  # Required

def hello_world():
    cred = DefaultAzureCredential()
    token = cred.get_token(scope)
    r = redis.Redis(host=host,
                    port=port,
                    ssl=True,    # ssl connection is required.
                    username=user_name,
                    password=token.token,
                    decode_responses=True)
    r.set("Az:key1", "value1")
    t = r.get("Az:key1")
    print(t)

if __name__ == '__main__':
    hello_world()
```

##### Supported Token Credentials for Azure AD Authentication

> **Note:** The samples in this doc use the azure-identity library's `DefaultAzureCredential` to fetch Azure AD Access Token. The other supported `TokenCredential` implementations that can be used from azure-identity for Python are as follows:

Reviewer 3: ```suggestion
**Note:** The samples in this doc use the azure-identity library's `DefaultAzureCredential` to fetch an Azure AD access token. The other supported `TokenCredential` implementations that can be used from azure-identity are as follows:
```

- [Client Certificate Credential](https://aka.ms/azsdk/python/identity/certificatecredential)
- [Client Secret Credential](https://aka.ms/azsdk/python/identity/clientsecretcredential)
- [Managed Identity Credential](https://aka.ms/azsdk/python/identity/managedidentitycredential)
- [Username Password Credential](https://aka.ms/azsdk/python/identity/usernamepasswordcredential)
- [Azure CLI Credential](https://aka.ms/azsdk/python/identity/azclicredential)
- [Interactive Browser Credential](https://aka.ms/azsdk/python/identity/interactivebrowsercredential)
- [Device Code Credential](https://aka.ms/azsdk/python/identity/devicecodecredential)

#### Authenticate with Azure AD: handle re-authentication

> This sample is intended to assist in authenticating with Azure AD via redis client library. It focuses on displaying the logic required to fetch an Azure AD access token and to use it as password when setting up the redis instance. It also shows how to recreate and authenticate the redis instance when its connection is broken in error/exception scenarios.

Reviewer 3: ```suggestion
This sample is intended to assist in authenticating with Azure AD via the redis client library. It focuses on displaying the logic required to fetch an Azure AD access token and to use it as password when setting up the redis instance. It also shows how to recreate and authenticate the redis instance when its connection is broken in error/exception scenarios.
```

    r = redis.Redis(host=host,
                    port=port,
                    ssl=True,    # ssl connection is required.
                    username=user_name,
                    password=token.token,
                    decode_responses=True)
    max_retry = 3
    for index in range(max_retry):
        try:
            r.set("Az:key1", "value1")
            t = r.get("Az:key1")
            print(t)
            break
        except redis.ConnectionError:
            _LOGGER.info("Connection lost. Reconnecting.")
            token = cred.get_token(scope)
            r = redis.Redis(host=host,
                            port=port,
                            ssl=True,    # ssl connection is required.
                            username=user_name,
                            password=token.token,
                            decode_responses=True)
        except Exception:
            _LOGGER.info("Unknown failures.")
            break

if __name__ == '__main__':
    re_authentication()
```

#### Troubleshooting

##### Invalid Username Password Pair Error

In this error scenario, the username provided and the access token used as password are not compatible.

> To mitigate this error, ensure that:
> 
> - On Portal, Under your Redis Cache Resource -> RBAC Rules, you've assigned the required role to your user/service principal identity.
> - On Portal, Under your Redis Cache Resource -> Advanced settings -> AAD Access Authorization box is checked/enabled, if not enable it and press the Save button.

Reviewer 3: ```suggestion
To mitigate this error, navigate to your Azure Cache for Redis resource in the Azure portal. Confirm that:

- In **RBAC Rules**, you've assigned the required role to your user/service principal identity.
- In **Advanced settings**, the **AAD access authorization** box is selected. If not, select it and select the **Save** button.
```

##### Permissions not granted / NOPERM Error

In this error scenario, the authentication was successful, but your registered user/service principal is not granted the RBAC permission to perform the action.

> To mitigate this error, ensure that:
> 
> - On Portal, Under your Redis Cache Resource -> RBAC Rules, you've assigned the appropriate role (Owner, Contributor, Reader) to your user/service principal identity.
> - In the event you're using a custom role, then ensure the permissions granted under your custom role include the one required for your target action.

Reviewer 3: ```suggestion
To mitigate this error, navigate to your Azure Cache for Redis resource in the Azure portal. Confirm that:

- In **RBAC Rules**, you've assigned the appropriate role (Owner, Contributor, Reader) to your user/service principal identity.
- In the event you're using a custom role, ensure the permissions granted under your custom role include the one required for your target action.
```
